### PR TITLE
New training

### PR DESCRIPTION
**Pull request by Ben-Graham**

### CODE REVIEW

File: Basic-Recruit-Training/README.adoc [hunk 1]
```diff
@@ -1,8 +1,86 @@
...
+
+----
+//function prototype
+void foo(int &a);
+
+int main()
+{
+    int x = 1;
+    int* p = &x;
+    foo(*p);
```
Comment (louisj381) [inline Basic-Recruit-Training/README.adoc]: wouldn't you just want to pass "p" into the argument? Since the function accepts the address of the function, not the value?

File: Basic-Recruit-Training/README.adoc [hunk 2]
```diff
@@ -1,8 +1,86 @@
...
+p = &x;
+----
+
+Printing the value of p will print the address of x.
+This will look something like 0x7ffc9b657c7c.
+The 0x is notation to show that the following number is in hexadecimal.
+This means that the value for x is stored on your machine at the memory location 0x7ffc9b657c7c.
+The real power of pointers comes from the use of the dereferencing operator, which is also an asterisk.
+Lets look back at the code example.
+
```
Comment (louisj381) [inline Basic-Recruit-Training/README.adoc]: In this example, I think it would be useful to illustrate the difference between * during declaration "int *p" and dereferencing operator when assigning a variable

File: Basic-Recruit-Training/README.adoc [hunk 3]
```diff
@@ -1,8 +1,86 @@
...
+In order to assign our pointer p to the address of a variable we will need to use the address operator, &.
+Take note of the following example:
+
+----
+int x;
+int *p;
+
+x = 3;
+p = &x;
+----
```
Comment (louisj381) [inline Basic-Recruit-Training/README.adoc]: I like that you go into dereferencing operator after the address, however it might be useful to introduce it in the first example(*p = ?), or pose the question "what do you think *p represents". That way the reader starts thinking about what it means as you transition into the dereferencing operator.

File: Basic-Recruit-Training/README.adoc [hunk 4]
```diff
@@ -65,7 +71,7 @@ int main()
 
 void foo(int &a)
 {
-    //some cool code
+    //some code
```
Comment (bill-luu) [inline Basic-Recruit-Training/README.adoc]: lol

File: Basic-Recruit-Training/README.adoc [hunk 5]
```diff
@@ -24,12 +34,11 @@ x = 3;
...
 ----
 
+Now the memory address of x will be stored in p.
 Printing the value of p will print the address of x.
 This will look something like 0x7ffc9b657c7c.
-The 0x is notation to show that the following number is in hexadecimal.
 This means that the value for x is stored on your machine at the memory location 0x7ffc9b657c7c.
-The real power of pointers comes from the use of the dereferencing operator, which is also an asterisk.
-Lets look back at the code example.
+Next lets look at the dereferencing operator.
```
Comment (bill-luu) [inline Basic-Recruit-Training/README.adoc]: `operator: `

File: Basic-Recruit-Training/README.adoc [hunk 6]
```diff
@@ -1,20 +1,30 @@
...
 
-For example:
+First lets look at a pointer variable.
 
 ----
 int * p;
 ----
 
-Will create a variable p, which is a pointer that will be able to point to a variable of type int.
+A pointer is a very similar to other primitive data types such as int.
```
Comment (bill-luu) [inline Basic-Recruit-Training/README.adoc]: Grammar. Reword to `A pointer is very similar to other ....`

File: Basic-Recruit-Training/README.adoc [hunk 7]
```diff
@@ -1,20 +1,30 @@
...
+x = 0;
+int *p;
+p = &x;
+*p = 6;
+
+cout << x << endl; //this prints the number 6
+----
 
-For example:
+First lets look at a pointer variable.
```
Comment (bill-luu) [inline Basic-Recruit-Training/README.adoc]: `variable:`

File: Basic-Recruit-Training/README.adoc [hunk 8]
```diff
@@ -1,8 +1,92 @@
...
+
+First lets look at a pointer variable.
+
+----
+int * p;
+----
+
+A pointer is a very similar to other primitive data types such as int.
+Instead of containing a value a pointer will instead contain a memory address.
+This address is the location in memory of another variable.
```
Comment (bill-luu) [inline Basic-Recruit-Training/README.adoc]: I would reword the previous two lines to: `Instead of containing values, pointers contain memory addresses to the location of another variable.`

File: Basic-Recruit-Training/README.adoc [hunk 9]
```diff
@@ -1,8 +1,92 @@
...
+int *p;
+int y;
+
+x = 3;
+p = &x;
+
+y = *p;
+----
+
+Printing the value of y will print 3.
```
Comment (bill-luu) [inline Basic-Recruit-Training/README.adoc]: Add a `cout` to the code like the first snippet instead of stating it in the text.

File: Basic-Recruit-Training/README.adoc [hunk 10]
```diff
@@ -1,8 +1,92 @@
...
+----
+int x;
+int *p;
+
+x = 3;
+p = &x;
+----
+
+Now the memory address of x will be stored in p.
+Printing the value of p will print the address of x.
```
Comment (bill-luu) [inline Basic-Recruit-Training/README.adoc]: Add a `cout` to the code like the first snippet instead of stating it in the text.

File: Basic-Recruit-Training/README.adoc [hunk 11]
```diff
@@ -1,8 +1,92 @@
...
+int y;
+
+x = 3;
+p = &x;
+
+y = *p;
+----
+
+Printing the value of y will print 3.
+We use the dereferencing operator to get the value stored at the address pointed to by the pointer.
```
Comment (bill-luu) [inline Basic-Recruit-Training/README.adoc]: We should mention that the address points to the address of the first byte of the object. Also that since this is an int pointer, dereferencing the value gives the value of the 4 bytes it points to.

File: Basic-Recruit-Training/README.adoc [hunk 12]
```diff
@@ -1,8 +1,92 @@
...
+void foo(int &a);
+
+int main()
+{
+    int x = 1;
+    int* p = &x;
+    foo(*p);
+}
+
+void foo(int &a)
```
Comment (bill-luu) [inline Basic-Recruit-Training/README.adoc]: Hey isn't this passing by reference? Passing a pointer would be `int *a`

File: Basic-Recruit-Training/pointers/main.cpp [hunk 13]
```diff
@@ -0,0 +1,58 @@
+/*
+First read through main() and makr predictions on what the outpout of each variable will be where indicated.
```
Comment (bill-luu) [inline Basic-Recruit-Training/pointers/main.cpp]: `makr`

File: Basic-Recruit-Training/README.adoc [hunk 14]
```diff
@@ -1,8 +1,92 @@
-Basic-Recruit-Training
------------------------
+= Pointers
```
Comment (bill-luu) [inline Basic-Recruit-Training/README.adoc]: I think it would make sense to split this README into two; one in the `Person` exercise and one for the `Pointer` exercise.

This readme would then be an overview of the entire Basic-Recruit-Training` - one that talks about the content of each folder.

File: Basic-Recruit-Training/README.adoc (hunk 12 above)
Comment (Ben-Graham) [reply Basic-Recruit-Training/README.adoc]: It works correctly with that code, but changing it to *a makes it more understandable

File: Basic-Recruit-Training/README.adoc [hunk 15]
```diff
@@ -1,92 +1,8 @@
-= Pointers
+= Basic Recruit Training
 
-One of the most important concepts in C/C++ programming is pointers.
-We will look at an example chunk of code that uses pointers to change the value in a variable before exploring pointers more in depth.
+There are two components to this training.
+== Person
+The directory person contains an exercize for a basic introduction to c++.
```
Comment (bill-luu) [inline Basic-Recruit-Training/README.adoc]: `exercize` -> `exercise`

File: Basic-Recruit-Training/README.adoc [hunk 16]
```diff
@@ -1,92 +1,8 @@
...
-
-In this exercise, you must create the `Person.cpp` file and implement the definition of all of the functions within `Person.h`.
-In addition, you must complete the definitions of `mySwap()` and `myIncrement()` and implement them in `main.cpp`.
-You are not permitted to modify anything in `Person.h` or any of the function headers.
-The program will verify that you have correctly completed the exercise.
-
-Both projects were designed for the Code::Blocks IDE, download here from http://www.codeblocks.org/downloads/binaries.
-Make sure that the binary you download contains the mingw compiler, i.e `codeblocks-16.01mingw-setup.exe`
+== Pointer
+The directory pointer contains an exercize for practice with a more advanced c++ concept called pointers.
```
Comment (bill-luu) [inline Basic-Recruit-Training/README.adoc]: `exercize` -> `exercise`

File: Basic-Recruit-Training/pointers/README.adoc [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+
+By passing the pointer p into the function instead of x we can change the value that the pointer is pointing to.
+
+Pointers are also very useful when working with data structures such as arrays.
+In the case of an array a pointer will point to the first byte of the array.
+If the array is of type int, which is 4 bytes, then the address of each array index will be 4 bytes greater then the previous one.
+
+Now move on to the pointer exercise in main.cpp and follow the directions in there
+
+
```
Comment (bill-luu) [inline Basic-Recruit-Training/pointers/README.adoc]: Let's add some links for further reading:
[http://www.cplusplus.com/doc/tutorial/pointers/](url)
